detach the volume when pod is terminated

gnufied opened the pull request:
When pods are terminated we should detach the volume. 

Fixes https://github.com/kubernetes/kubernetes/issues/45191

**Release note**:
```
Detach the volume when pods are terminated.  
```

gnufied committed:
detach the volume when pod is terminated

Make sure volume is detached when pod is terminated because
of any reason and not deleted from api server.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45286)
<!-- Reviewable:end -->

wongma7 commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -395,8 +401,23 @@ func (adc *attachDetachController) GetDesiredStateOfWorld() cache.DesiredStateOf
 }
 
 func (adc *attachDetachController) podUpdate(oldObj, newObj interface{}) {
```
podUpdate still only happens once every 12 hours though right?

I think we have to do it in one of the reconciler loops

gnufied replied:
I would assume that, `podUpdate` should get called when pod's status for example, changes from `Running` to `Complete` - not every 12 hours.

gnufied replied:
Also, regardless of this event though - I have also updated DSWP to remove the pod from DSW if pod is terminated.

wongma7 replied:
oh yeah...brain fart my b 🥇

luxas reviewed (commented):
Thanks for this PR!

General comment without having context on this particular feature -- is there any way we could implement this without adding yet another flag? We really need to ask ourselves that question, otherwise we'll end up with thousands of flags. Just something to keep in mind ;)

luxas commented:
/unassign

childsb commented:
@luxas the reason for the flag is to preserver backwards behavior, which in some cases may be desirable.

rootfs commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -233,6 +236,9 @@ type attachDetachController struct {
 	// recorder is used to record events in the API server
 	recorder record.EventRecorder
+
+	// keep pod volumes attached for terminated pods
```
whether keep pod volumes

rootfs commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -395,8 +401,23 @@ func (adc *attachDetachController) GetDesiredStateOfWorld() cache.DesiredStateOf
+		return
+	}
+
+	util.ProcessPodVolumes(pod, true, /* addVolumes */
```
nit: make a boolean variable instead of calling `ProcessPodVolumes` twice with mostly the same parameters.

gnufied replied:
ack

rootfs commented on pkg/controller/volume/attachdetach/attach_detach_controller_test.go:
```diff
@@ -46,7 +46,8 @@ func Test_NewAttachDetachController_Positive(t *testing.T) {
 		false,
-		time.Second*5)
+		time.Second*5,
+		false /*keep pod volumes attached for terminated pods */)
```
whether keep pod volumes...

gnufied replied:
You mean change wording to `/* whether keep pod volumes */` ?

rootfs commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -211,6 +212,7 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet, allControllers []string, disabled
 	fs.Int32Var(&s.LargeClusterSizeThreshold, "large-cluster-size-threshold", 50, "Number of nodes from which NodeController treats the cluster as large for the eviction logic purposes. --secondary-node-eviction-rate is implicitly overridden to 0 for clusters this size or smaller.")
 	fs.Float32Var(&s.UnhealthyZoneThreshold, "unhealthy-zone-threshold", 0.55, "Fraction of Nodes in a zone which needs to be not Ready (minimum 3) for zone to be treated as unhealthy. ")
 	fs.BoolVar(&s.DisableAttachDetachReconcilerSync, "disable-attach-detach-reconcile-sync", false, "Disable volume attach detach reconciler sync. Disabling this may cause volumes to be mismatched with pods. Use wisely.")
+	fs.BoolVar(&s.KeepTerminatedPodVolumes, "keep-terminated-pod-volumes", false, "Keep terminated pod volumes attached to the node after the pod terminates.  Can be useful for debugging volume related issues.")
```
-> keep volumes attached to node ?

rootfs commented on pkg/apis/componentconfig/types.go:
```diff
@@ -896,6 +896,10 @@ type KubeControllerManagerConfiguration struct {
 	// through the API server proxy.
 	HorizontalPodAutoscalerUseRESTClients bool
+
+	// KeepTerminatedPodVolumes causes terminated pod volumes attached to the node after the pod terminates.
```
=> KeepTerminatedPodVolumes keeps volumes attached to node after the pod terminates?

saad-ali commented:
Largely looks fine to me. A couple of concerns:
1. This makes the detach routine even more aggressive. I'm wary of doing that unless it is absolutely necessary since bugs here can result in a disk being pulled from a volume in use (and potential data loss).
2. `keep-terminated-pod-volumes` is a flag on `kubelet`. You are adding it to the `kube-controller-manager`. How do they stay in sync? This is why flags are a poor way to configure k8s components, but I guess that's beyond the scope of this PR. At the very least please update the existing references to also set the new reference (e.g. `hack/local-up-cluster.sh`).

Will leave it to @jingxu97 for a more in depth review.

gnufied commented:
For concern-1, I think this is needed because Completed, Failed or Terminated pods currently leave the volume attached to the node and currently `kubectl get pods` for example doesn't even show completed pods, as a result we have users thinking they can reuse the PVC in other pod. But obviously that doesn't work.  We have had many users struggling with this.  cc @eparis 

For concern-2, yeah it is bit iffy. I am hoping that most users would run with default and hence kubelet and controller flags will be in sync. But I am also thinking of improving documentation of the flag in controller and mention corresponding flag in kubelet, so as users know about both of them. ack about updating references in `./hack/**`

jingxu97 commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -395,8 +401,23 @@ func (adc *attachDetachController) GetDesiredStateOfWorld() cache.DesiredStateOf
+		addPodFlag = false
+	}
+
+	util.ProcessPodVolumes(pod, addPodFlag, /* addVolumes */
```
I am thinking to be consistent with kubelet, for podAdd function, we can also check whether the pod is already terminated or not. It might be necessary if podAdd means pod would never be in terminated state...

gnufied replied:
We do check that right? On line#415 (https://github.com/kubernetes/kubernetes/pull/45286/files#diff-72e66cba2ced392dc0f9cd10c424d6d0R415 ) 

I do not think there is any chance we are going to detach volumes still mounted and being used inside pod.

jingxu97 replied:
but that is for podUpdate, not podAdd?

gnufied replied:
fixed

eparis commented:
We could make kubelet set "keep-terminated-pod-volumes" as a node annotation (similar to how we do master-attach-detach) then use that? I hate it, but we can if someone says we have to. I want no new flags if there is any way to do that. Not everything needs to be a choice.

I do question if setting keep-terminated-pod-volumes 'per-node' is/was really appropriate or if it should be a cluster wide things.

rootfs commented:
also see #45346

sjenning commented:
@eparis yeah, at the time I introduce that code and the new flag, it was geared toward freeing up emptydirs on pod termination.  The detach-attach part of the PVs that might be associated with those mounts were outside the discussion at the time.

Like Saad, I do worry about keeping the flags in sync.  If they are not, the attach-detach controller could end up force detaching mounted volumes and maybe causing corruption if the data previously written by the terminated pod is not synced before the forced detach.

However, if you can sync the data to volumes upon pod termination, I guess the worst that could happen is that your volumes are not their for debugging even if you set `--keep-terminated-pod-volumes` on the kubelet.

gnufied commented:
Spoke with @jingxu97 a bit offline about k8s force detaching mounted volumes. It appears that this mechanism was mostly introduced to recover volumes from unresponsive nodes or kubelets. But I think that - it may actually be unnecessary.  

Usually terminating problematic node automatically detaches volumes from it and force detaching volumes *is* always dangerous regardless of this PR.  Even without this PR - We have had problem where volumes used in pods were getting detached. https://github.com/kubernetes/kubernetes/issues/19602

jsafrane commented on cmd/kubelet/app/options/options.go:
```diff
@@ -293,7 +293,7 @@ func (c *kubeletConfiguration) addFlags(fs *pflag.FlagSet) {
 	fs.Int32Var(&c.PodsPerCore, "pods-per-core", c.PodsPerCore, "Number of Pods per core that can run on this Kubelet. The total number of Pods on this Kubelet cannot exceed max-pods, so max-pods will be used if this calculation results in a larger number of Pods allowed on the Kubelet. A value of 0 disables this limit.")
 	fs.BoolVar(&c.ProtectKernelDefaults, "protect-kernel-defaults", c.ProtectKernelDefaults, "Default kubelet behaviour for kernel tuning. If set, kubelet errors if any of kernel tunables is different than kubelet defaults.")
-	fs.BoolVar(&c.KeepTerminatedPodVolumes, "keep-terminated-pod-volumes", c.KeepTerminatedPodVolumes, "Keep terminated pod volumes mounted to the node after the pod terminates.  Can be useful for debugging volume related issues.")
+	fs.BoolVar(&c.KeepTerminatedPodVolumes, "keep-terminated-pod-volumes", c.KeepTerminatedPodVolumes, "Keep terminated pod volumes mounted to the node after the pod terminates.  Can be useful for debugging volume related issues. If setting this value to true - make sure keep-terminated-pod-volumes options in controller is also set to true to avoid detaching mounted volumes.")
```
I see a new kubelet option, but I don't see it's used anywhere in kubelet...

wongma7 replied:
It's not new, it happened to exist already so he updated its description

jsafrane replied:
oops :innocent:

jsafrane commented:
I'm not sure I fully understand the problem, but if a volume needs to attach/detach, then it's probably persistent enough to do any emergency recovery even when the volume is detached automatically by kubelet/controller, right? We need to keep only emptyDir and similar ephemeral storage. Dunno how easy/hard/messy would it be to do such distinction (half of volumes are unmounted/detached, second half is kept... sounds scary)

jsafrane commented:
> Even without this PR - We have had problem where volumes used in pods were getting detached. #19602

I think we fixed these bugs by introducing attach/detach controller. I may be wrong though, it's a long time ago.

jsafrane commented:
Umm, thinking out loud, if kubelet has `--keep-terminated-pod-volumes` flag, it won't unmount any volumes and thus it won't remove them from `Node.Status.VolumesInUse`, so attach/detach controller won't ever detach them, right? Unless I am missing something, we can be quite sure that we can detach a volume if it *isn't* in `Node.Status.VolumesInUse` and its pod is `Terminated`. And we don't need any flags or annotations.

eparis commented:
I know this isn't the right PR, but the right people are here to scream no. Anyone have a problem if we mark `keep-terminated-pod-volumes` deprecated and move towards killing it off? It doesn't really make sense as a node flag. I'm not sure what the value of it at all is...

jingxu97 commented:
@jsafrane  
"if kubelet has --keep-terminated-pod-volumes flag, it won't unmount any volumes and thus it won't remove them from Node.Status.VolumesInUse, so attach/detach controller won't ever detach them, right?"

Because we have a 6 minute timeout, attach/detach controller will detach volume after 6 minutes no matter it is still in Node.Status.VolumesInUse or not. I think we might need to change this behavior with care. 

The problem this PR addresses it that when the flag --keep-terminated-pod-volumes is not set (by default), controller should detach terminated pods. Currently, controller will not detach terminated pods unless user delete them from the api server, however, kubelet will unmount them.

gnufied commented:
Yep @jsafrane jing is right, the volumes are detached after 6 minute timeout. Regardless of the fact that, if they are mounted or not.  Now as it happens, AWS is *only* cloudprovider that afaik doesn't allow detach of mounted volumes - so this mechanism doesn't work there.

@jingxu97 Do we have any major concern that I am missing here - apart from that, if *somehow* user has enabled `keep-terminated-pod-volumes` in kubelet for debugging some volume failure and did not enable it in controller, the controller may go ahead and detach it?

If user is setting that flag only in kubelet, I have updated documentation to remind user to set it in controller too.

But we may want to revisit the flag in future. But I do not think is in the scope of this PR. As it exists today, users with completed/failed pods are facing problems reusing their PVCs and that we should fix.

jingxu97 commented:
@gnufied @saad-ali I am thinking to remove 6 minutes timeout detach logic. In this way, if user forgot to set the controller flag, it will not detach while it is still mounted. (maybe flag for controller is also not necessary). 

In case of kubelet not responsive (I assume this normally happens when the node has serious problem), since controller does not know what is going on the node, it is not safe to just issue a detach. User have to manually check the node and determine the actions.

gnufied commented:
I agree. @eparis was suggested the same. This would remove need for having this flag in controller.

saad-ali commented:
> eparis commented 4 hours ago:
>> Anyone have a problem if we mark keep-terminated-pod-volumes deprecated and move towards killing it off? It doesn't really make sense as a node flag. I'm not sure what the value of it at all is...

I'm ok with this.

> jingxu97 commented 2 hours ago:
>> @gnufied @saad-ali I am thinking to remove 6 minutes timeout detach logic. In this way, if user forgot to set the controller flag, it will not detach while it is still mounted. (maybe flag for controller is also not necessary).
>> 
>> In case of kubelet not responsive (I assume this normally happens when the node has serious problem), since controller does not know what is going on the node, it is not safe to just issue a detach. User have to manually check the node and determine the actions.

I wouldn't remove the logic all together. Maybe increase the limit if you're worried. But I'd like to see evidence that this has caused issues either way.

gnufied commented:
> I'm ok with this.

Can you elaborate? I am thinking we should drop the controller flag I added here and move towards deprecating/removing kubelet flag. 

> I wouldn't remove the logic all together. Maybe increase the limit if you're worried. But I'd like to see evidence that this has caused issues either way.

The reason I was thinking of removing it is, not because it is causing issues but I don't think it helps and there is quite a lot of code to facilitate this mechanism.  We can delete the whole `DetachRequestTime` thingy if we decide to stop force detaching the volumes.

gnufied commented:
@saad-ali By elaborate I meant , if we are going to deprecate the node flag how do we proceed with this PR. Do you agree that?

1. We drop the corresponding flag that I introduced in the controller?
2. We keep rest of this PR same as it is and we consider behavior of force detach in another PR/issue?

saad-ali commented:
> The reason I was thinking of removing it is, not because it is causing issues but I don't think it helps and there is quite a lot of code to facilitate this mechanism. We can delete the whole DetachRequestTime thingy if we decide to stop force detaching the volumes.

It may actually be helping in some cases but going unnoticed, so I'd prefer not to remove it without sufficient. I'm afraid pulling doing so may result in a lot of "my volume is not getting attached" support calls.

> We drop the corresponding flag that I introduced in the controller?

That's fine. Just make sure to add a comment on the flag (in the code, etc.) indicating the new breaking behavior and open an issue to deprecate the flag.

gnufied committed:
Add Keepterminatedpodvolumes as a annotation on node

and lets make sure that controller respects it
and doesn't detaches mounted volumes.

gnufied commented:
@saad-ali @jingxu97 I spoke with @jsafrane  offline a bit and I have pushed a commit that adds `volume.kubernetes.io/keep-terminated-pod-volumes` annotation to node object if corresponding flag has been set in the kubelet. If that flag is set for node, we don't detach volumes for terminated pods.

I think this is best middleground we have and this also makes sure none of existing behavior breaks. PTAL.

gnufied commented:
@k8s-bot kops aws e2e test this

gnufied commented:
Created an issue to deprecate `keep-terminated-pod-volumes` flag on kubelet - https://github.com/kubernetes/kubernetes/issues/45615

saad-ali commented on pkg/controller/volume/attachdetach/cache/desired_state_of_world.go:
```diff
@@ -313,6 +324,30 @@ func (dsw *desiredStateOfWorld) VolumeExists(
+// return values:
+//    true  -- if volume needs to be attached
+//    false -- if volume needs to be detached
+func (dsw *desiredStateOfWorld) DetermineVolumeAction(pod *v1.Pod, defaultValue bool) bool {
```
This is a really funky method to add to `desired_state_of_the_world.go`--it blurs the boundary of what this class is responsible for. Think about what the responsibilities of this code is and if it makes sense for this logic to live here, and use that to refactor this. If you want my thoughts on how this should be factored let me know.

gnufied replied:
yeah I was debating between moving this to util vs keeping in dsw. The problem is - dswp in this case holds `managedNode` and is a unexported member of `cache` package and hence can't be accessed directly. 

In this case dws along with pod holds the truth about volume action (attach or detach) needed on the pod.  I am okay moving this to util, but we will have to export the `managedNode` type from `cache` package or something similar.

Having said that - what are your thoughts on this?

gnufied replied:
We do need some function in dsw to get node state out of it, in one form or another - right?

saad-ali replied:
This method is called in three places.
* `desired_state_of_world_populator.go:findAndRemoveDeletedPods()`
  * In this case the pods being operated on are fetched via a `dswp.desiredStateOfWorld.GetPodToAdd()` call so you can assume that the node is managed--no need to check again, just call `volumehelper.IsPodTerminated()`.
* `attach_detach_controller.go:podAdd()` and `attach_detach_controller.podUpdate()`
  * Both call this method first and then `util.ProcessPodVolumes(...)`.
  * `util.ProcessPodVolumes(...)` already calls `desiredStateOfWorld.NodeExists(...)` to verify the node is managed. So you don't need to do the check, `volumehelper.IsPodTerminated(...)` should be sufficient to determine if the volume should be added or deleted.

Additionally I would change this block of code:

```GO
if volumehelper.IsPodTerminated(pod, pod.Status) {
	util.ProcessPodVolumes(pod, false, /* addVolumes */
		adc.desiredStateOfWorld, &adc.volumePluginMgr, adc.pvcLister, adc.pvLister)
} else {
	util.ProcessPodVolumes(pod, true, /* addVolumes */
		adc.desiredStateOfWorld, &adc.volumePluginMgr, adc.pvcLister, adc.pvLister)
}
```
to 
```GO
addVolumes := !volumehelper.IsPodTerminated(pod, pod.Status)

util.ProcessPodVolumes(pod, addVolumes, adc.desiredStateOfWorld, &adc.volumePluginMgr, adc.pvcLister, adc.pvLister)
```

That will allow you to get rid of `desired_state_of_world.go:DetermineVolumeAction(...)` altogether.

saad-ali replied:
To handle the new `keepTerminatedPodVolumes` annotation you can add a new `keepTerminatedPodVolumes` field per node to the DSW cache, populate it on node creation, and add a new `GetKeepTerminatedPodVolumesForNode(node)` method to expose it. Then wherever you call `volumehelper.IsPodTerminated(...)` first call the new method. This is a better approach because it ensures that `desired_state_of_the_world.go` remains "just a cache"--meaning it just stores state info and allows thread-safe access to it, it does not implement business logic of "what to do".

gnufied replied:
I have removed that logic from desired_state_of_world and moved it to `util.go`. PTAL

saad-ali reviewed (dismissed):
Fine with the direction, some comments on the factoring of the new code.

gnufied commented:
@k8s-bot unit test this

jsafrane commented on pkg/controller/volume/attachdetach/util/util.go:
```diff
@@ -159,6 +159,23 @@ func getPVSpecFromCache(name string, pvcReadOnly bool, expectedClaimUID types.UI
 
+// DetermineVolumeAction returns true if volume and pod needs to be added to dswp
+// and it returns false if volume and pod needs to be removed from dswp
+func DetermineVolumeAction(pod *v1.Pod, desiredStateOfWorld cache.DesiredStateOfWorld, defaultValue bool) bool {
```
`DetermineVolumeAction` is not a good name, we already know the action, we need to update it for terminated pods. Also `defaultValue` -> `defaultAction` (or `action`).

gnufied replied:
What you mean - we already know the action? The volume action (attach or detach) is decided not only by state of the pod (terminated or not) but also if user has asked kubelet to keep terminated volumes around. 

I am happy to change function name though. The only thing I don't like about this function is - it returns a boolean variable whereas it sounds like it is going to return name of the action (`attach` or `detach`). So this function is bit like a gate/latch - that determines which way the code will flow.

jsafrane replied:
> The volume action (attach or detach) is decided not only by state of the pod (terminated or not) but also if user has asked kubelet to keep terminated volumes around.

IMO, the most important decision is done before calling the function and is passed as `defaultValue`. That's why the name seems misleading to me - `DetermineVolumeAction` implies that whole decision is done here. Anyway, it's just a function name, DetermineVolumeAction will do.

jsafrane commented on pkg/controller/volume/attachdetach/util/util.go:
```diff
@@ -159,6 +159,23 @@ func getPVSpecFromCache(name string, pvcReadOnly bool, expectedClaimUID types.UI
+// and it returns false if volume and pod needs to be removed from dswp
+func DetermineVolumeAction(pod *v1.Pod, desiredStateOfWorld cache.DesiredStateOfWorld, defaultValue bool) bool {
+	if pod == nil || len(pod.Spec.Volumes) <= 0 {
+		return defaultValue
```
This implies this function could return `true` if pod is nil or has no volumes... Is this intended?

gnufied replied:
In all cases if pod or spec has no volumes - the underlying operation is going to be a no-op. So in reality we don't really care what value we return from here.

But rather than hardcoding `true` or `false` from such branches - we let the caller or the function decide what he expects. For example - the default action to take in case of `podAdd` event is to attach the volume. So we let caller of `DetermineVolumeAction` supply a default action.  I implemented it this way to make sure - we are backward compatible as much as possible - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/attach_detach_controller.go#L388

As you can see from above line - the default behaviour on `podAdd` is `true`.

jsafrane replied:
ack, ignore my comment then

jsafrane reviewed (commented):
I generally like the PR, I only have comments about function names and other silly stuff.

jsafrane commented:
lgtm from my side (I am not assignee though)

eparis commented:
/approve

jsafrane commented:
/approve

deads2k commented:
/approve

derekwaynecarr commented on hack/local-up-cluster.sh:
```diff
@@ -208,6 +208,7 @@ CPU_CFS_QUOTA=${CPU_CFS_QUOTA:-true}
 ENABLE_HOSTPATH_PROVISIONER=${ENABLE_HOSTPATH_PROVISIONER:-"false"}
 CLAIM_BINDER_SYNC_PERIOD=${CLAIM_BINDER_SYNC_PERIOD:-"15s"} # current k8s default
 ENABLE_CONTROLLER_ATTACH_DETACH=${ENABLE_CONTROLLER_ATTACH_DETACH:-"true"} # current default
+KEEP_TERMINATED_POD_VOLUMES=${KEEP_TERMINATED_POD_VOLUMES:-"true"}
```
why is this defaulting to true?  i don't think you should do this change.

derekwaynecarr replied:
actually, i see below it was defaulting to true previously... i think that was a mistake... can we make that false?

eparis replied:
@derekwaynecarr I think that change should be a part of https://github.com/kubernetes/kubernetes/issues/45615

derekwaynecarr reviewed (changes requested):
can we update hack/local-up to default to false, otherwise, this looks fine to me.

derekwaynecarr reviewed (approved):
per https://github.com/kubernetes/kubernetes/issues/45615. i am fine not waiting for local up to change.

derekwaynecarr commented:
/approve

childsb commented:
/lgtm

gnufied commented:
@k8s-bot bazel test this.

saad-ali commented:
Taking a quick final pass

saad-ali commented on pkg/controller/volume/attachdetach/cache/desired_state_of_world.go:
```diff
@@ -144,6 +148,9 @@ type nodeManaged struct {
 	// the value is a pod object containing more information about the volume.
 	volumesToAttach map[v1.UniqueVolumeName]volumeToAttach
+
+	// keepTerminatedPodVolumes
```
nit: More descriptive comment please!

saad-ali commented on pkg/controller/volume/attachdetach/cache/desired_state_of_world.go:
```diff
@@ -46,7 +46,7 @@ type DesiredStateOfWorld interface {
 	// detach controller.
 	// If the node already exists this is a no-op.
-	AddNode(nodeName k8stypes.NodeName)
+	AddNode(nodeName k8stypes.NodeName, keepTerminatedPodVolumes bool)
```
nit: Update comment to describe the new parameter. This is particularly important for interfaces so that callers and implementers know what the expectations are.

saad-ali reviewed (commented):
A few nits. Feel free to address them here and do a self-LGTM or in a follow up PR.

saad-ali commented:
/lgtm
/approve

gnufied commented:
Given rather long build times right now because of bazel problems(https://github.com/kubernetes/test-infra/issues/2708,  https://github.com/kubernetes/kubernetes/issues/45558) - I am going to make the comment fixes as part of #45615.

I will make a note for myself. thanks @saad-ali

gnufied commented:
@k8s-bot bazel test this

eparis commented:
@k8s-bot unit test this

gnufied commented:
@eparis this needs a small code change now, which I am pushing. Basically master has changed contract of `RunAMaster` now - https://github.com/kubernetes/kubernetes/blob/master/test/integration/volume/attach_detach_test.go#L80

Obviously tests I added still uses old signature and it broke. Likely the change that changed the method signature was added after everything was green.

gnufied commented:
I have pushed a fix for integration tests which were failing because change in `RunAMaster` signature. Also fixed comments by @saad-ali while at it.

gnufied commented:
/lgtm
